### PR TITLE
samtools_stats

### PR DESCRIPTION
**Pull request by cjuigne**

### CODE REVIEW

**Comment by cjuigne** [comment]
still not good, I forgot $schemas

File: samtools/samtools_stats.cwl [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    inputBinding:
+      prefix: -g
+
+arguments:
+  - prefix: --threads
+    valueFrom: $(runtime.cores)  
+    
+# -X
+#     If this option is set, it will allows user to specify customized index file location(s) if the data folder does not contain any index file. Example usage: samtools stats [options] -X /data_folder/data.bam /index_folder/data.bai chrM:1-10
+
```
Comment (hellymac) [inline samtools/samtools_stats.cwl:140]: $schemas:
  - 'http://edamontology.org/EDAM_1.18.owl'

File: samtools/samtools_stats.cwl [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+- class: DockerRequirement
+  dockerPull: biocontainers/samtools:v1.7.0_cv3
+
+inputs:
+  input_file:
+    type: File
+     format: 
+      - edam:format_2572  # BAM
+      - edam:format_2573  # SAM
+      - edam:format_3462  # CRAM
```
Comment (mr-c) [inline samtools/samtools_stats.cwl:16]: Have you tested this? `samtools stats --help` says "The program collects statistics from BAM files."

File: samtools/samtools_stats.cwl [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+
+outputs:
+  stats:
+    type: File
+    outputBinding:
+      glob: $(inputs.input_file.nameroot).stats.txt
+stdout: $(inputs.input_file.nameroot).stats.txt
+
+$schemas:
+  - 'http://edamontology.org/EDAM_1.18.owl'
```
Comment (mr-c) [inline samtools/samtools_stats.cwl:151]:
```suggestion
$namespaces: { edam: http://edamontology.org, iana: https://www.iana.org/assignments/media-types/ }
$schemas:
  - 'http://edamontology.org/EDAM_1.18.owl'
```

File: samtools/samtools_stats.cwl [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+    default: 0
+    doc: " STR|INT Required flag, 0 for unset. See also `samtools flags` [0] "
+    inputBinding:
+      prefix: -f
+
+  filtering_flag:
+    type:
+      - string
+      - int
+      - "null"|
```
Comment (mr-c) [inline samtools/samtools_stats.cwl]:
```suggestion
      - "null"
```

If that was from my original suggestion, my apologies!

File: samtools/samtools_stats.cwl [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+
+baseCommand: [samtools, stats]
+requirements:
+- class: DockerRequirement
+  dockerPull: biocontainers/samtools:v1.7.0_cv3
+
+inputs:
+  input_file:
+    type: File
+     format:
```
Comment (mr-c) [inline samtools/samtools_stats.cwl]:
```suggestion
  input_file:
    type: File
    format: 
```

File: samtools/samtools_stats.cwl [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+    type: boolean?
+    doc: "Exclude from statistics reads marked as duplicates"
+    inputBinding:
+      prefix: --remove_dups
+
+  required_flag:
+    type:
+      - string
+      - int
+      - "null"|
```
Comment (mr-c) [inline samtools/samtools_stats.cwl]:
```suggestion
      - "null"
```

File: samtools/samtools_stats.cwl [hunk 7]
```diff
@@ -0,0 +1,150 @@
+class: CommandLineTool
+cwlVersion: v1.0
+
```
Comment (mr-c) [inline samtools/samtools_stats.cwl]:
```suggestion
#!/usr/bin/env cwl-runner
cwlVersion: v1.0
class: CommandLineTool

```

File: samtools/samtools_stats.cwl (hunk 2 above)
Comment (cjuigne) [reply samtools/samtools_stats.cwl:16]: I tested on SAM and BAM files, not on CRAM file, but in the doc it is said "samtools stats [options] in.sam|in.bam|in.cram [region...] "

File: samtools/samtools_stats.cwl [hunk 8]
```diff
@@ -146,6 +146,6 @@ outputs:
       glob: $(inputs.input_file.nameroot).stats.txt
 stdout: $(inputs.input_file.nameroot).stats.txt
 
-$namespaces: { edam: http://edamontology.org, iana: https://www.iana.org/assignments/media-types/ }
+$namespaces: { edam: http://edamontology.org/, iana: https://www.iana.org/assignments/media-types/ }
```
Comment (mr-c) [inline samtools/samtools_stats.cwl:149]: Good catch, thanks!

File: samtools/samtools_stats.cwl (hunk 2 above)
Comment (mr-c) [reply samtools/samtools_stats.cwl:16]: I think you have a different samtools version than the one in the Docker container:
```
$ docker run -it biocontainers/samtools:v1.7.0_cv3 samtools stats --help
About: The program collects statistics from BAM files. The output can be visualized using plot-bamstats.
Usage: samtools stats [OPTIONS] file.bam
       samtools stats [OPTIONS] file.bam chr:from-to
```

Comment (cjuigne) [reply samtools/samtools_stats.cwl:16]: Oh, I don't know, I tested with this cwl file and this container on SAM and BAM

Comment (mr-c) [reply samtools/samtools_stats.cwl:16]: Lets keep all three for now. It seems likely that CRAM works if SAM & BAM do as well.

**Comment by mr-c** [comment]
Thanks again!

**Pull request merged**